pkg/sync/sync_context.go

```diff
@@ -688,9 +690,34 @@ func (sc *syncContext) getSyncTasks() (_ syncTasks, successful bool) {
 		task.liveObj = sc.liveObj(task.targetObj)
 	}
 
+	isRetriable := func(err error) bool {
+		if apierr.IsUnauthorized(err) {
+			return true
+		}
+		return false
+	}
+
+	serverResCache := make(map[schema.GroupVersionKind]*metav1.APIResource)
+
 	// check permissions
 	for _, task := range tasks {
-		serverRes, err := kube.ServerResourceForGroupVersionKind(sc.disco, task.groupVersionKind(), "get")
+
+		var serverRes *metav1.APIResource
+		var err error
+
+		if val, ok := serverResCache[task.groupVersionKind()]; ok {
+			serverRes = val
+			err = nil
+		} else {
+			retry.OnError(retry.DefaultRetry, isRetriable, func() error {
```

Comment:
I think we need an abort condition here, like a maximum retry count. Otherwise it'll retry unless the function succeeds (and in case of a hard authorization error, probably this would be forever).

Reply:
1. Agree with @jannfis. You can use the .Cap field in wait.Backoff{} to cap the duration of retry.

2. Why aren't we checking the error return value of retry.OnError()?

Reply:
@jannfis @jessesuen

The DefaultRetry is already providing the max retry count:

```golang
var DefaultRetry = wait.Backoff{
	Steps:    5,
	Duration: 10 * time.Millisecond,
	Factor:   1.0,
	Jitter:   0.1,
}
```

The unit tests are validating that the retry will do max 5 attempts and if it fails it will return the last error.

> Why aren't we checking the error return value of retry.OnError()?

The error is handled by the code outside of the retry call. I kept the same logic there to preserve the current behaviour. The retry is just invoking additional attempts if the returned error is `Unauthorized` as implemented in the `isRetryable` function:
```golang
	isRetryable := func(err error) bool {
		if apierr.IsUnauthorized(err) {
			return true
		}
		return false
	}
```

Reply:
@jannfis @jessesuen Added one more test-case to validate that retry is just invoked if returned error is `Unauthorized`

Reply:
Ah you're right! The default retry has a default limit. Thanks for clarifying that.